### ISSUE TITLE
Unstable behavior of entanglement entropy of eigenstates

### ISSUE THREAD

**Issue by Alehud** [weinbe58/QuSpin#437]
I have been testing Quspin on a 2D spin model with a 4-spin interaction. The setup is the following. The model has a spin-1/2 at each site of the square lattice. The Hamiltonian consists of a ring exchange term,
<img src="https://user-images.githubusercontent.com/35610798/127925940-6a294908-0b84-4816-b30d-61a6a0fe1ef4.png" width="300">
where the summation goes over all elementary square plaquettes, and i,j,k,l are indices going around the plaquette. I have taken a 3x3 lattice with periodic boundary conditions and tried to plot the von Neumann entanglement entropy of eigenstates vs their eigenenergy, S(E). The bipartition I have chosen is: two first columns on the left belong to subsystem A (6 sites), and the last column belongs to subsystem B (3 sites). The code is below.
```python
import numpy as np
from quspin.operators import hamiltonian
from quspin.basis import spin_basis_general
import matplotlib.pyplot as plt

Lx = 3
Ly = 3
K = 1.0
N_2d = Lx*Ly
s = np.arange(N_2d)
x = s % Lx
y = s // Lx
T_x = (x+1) % Lx + Lx*y
T_y = x + Lx*((y+1) % Ly)
T_x_y = (x+1) % Lx + Lx*((y+1) % Ly)

basis = spin_basis_general(N_2d, pauli=False)

K_list = [[K, i, T_x[i], T_y[i], T_x_y[i]] for i in range(N_2d)]
print(K_list)
static = [["+--+", K_list], ["-++-", K_list]]
dynamic = []

H = hamiltonian(static, dynamic, basis=basis, dtype=np.float64, check_symm=False)
E, V = H.eigh()
ent = basis.ent_entropy(V, sub_sys_A=(0,1,3,4,6,7), density=False, enforce_pure=True)["Sent_A"]

fig, ax = plt.subplots(figsize=(14, 8), tight_layout=True)
ax.set_ylabel('entanglement entropy, $S_{AB}$', fontsize=20)
ax.set_xlabel('eigenenergy, $E$', fontsize=20)
ax.scatter(E, ent, c='red', s=20, marker='o')
plt.show()
```

Now, when in line 25 I change `dtype=np.float64` into `dtype=np.float32`, the spectrum changes drastically (see two images below), which, according to my understanding, should not be the case.
<img src="https://user-images.githubusercontent.com/35610798/127927782-a46710c3-4972-4a99-a5ad-657e7563dff9.png" width="400"> <img src="https://user-images.githubusercontent.com/35610798/127927791-2811c3f7-5d13-4800-b9e4-55daf9febf91.png" width="400">

In fact, I have my own code that calculates entanglement entropy for this model and the spectrum I get is different from both previous pictures (see picture below).
<img src="https://user-images.githubusercontent.com/35610798/127926413-61417e49-7132-4fa6-9e9a-dac6744cd88c.png" width="500">

I suspect that there might be some kind of numerical instability in your code (visually it seems that "your" spectrum is a very smeared version of "my" spectrum). As a guess, I can assume that the error might originate from the fact that in this model the reduced density matrix quite often has zero eigenvalues, and since x\*ln(x) -> 0 at x->0, these zero eigenvalues should not contribute towards S. Maybe quspin tries to calculate the value of x\*ln(x) explicitly?

Let me know what you think. Thanks!

P.S.
 - OS: Windows 10
 - Additional context: More info on the model can be found in a recent paper where I studied it numerically.
https://arxiv.org/abs/2107.09690
It has subsystem symmetries (magnetization in each row and each column is conserved) and exhibits extensive fragmentation of the Hilbert space, i.e., symmetry blocks are further block-diagonalized into kinetically disconnected sectors (on the 3x3 lattice with p.b.c., however, there is no fragmentation; the lattice is too small). I don't really know if these properties have anything to do with the issue above.

**Comment by mgbukov** [comment]
thanks! I was able to reproduce the difference in the float32 vs float64 behavior, although I got different plots on mac osx. Hence, it must be some issue with the numerical stability indeed. 

[Sent_bug.txt](https://github.com/weinbe58/QuSpin/files/6921145/Sent_bug.txt)

I also tried using scipy's svd solver instead of numpy but I got the same result (w/o tweaking the solver optional args though). 

@Alehud can you post the code you use to compute the entanglement so we can use this as a reference in debugging.

**Comment by Alehud** [comment]
[issue.zip](https://github.com/weinbe58/QuSpin/files/6934162/issue.zip)
Extract all files in a folder and run main.py. Sorry for many files, I just took them out of a larger project. I tried to remove everything unrelated to this issue from the code.

My code calculates the spectrum separately in each block of the full Hilbert space (these blocks don't necessarily have symmetry associated with them, hence the name "Hilbert space fragmentation", which you can read more about in my paper mentioned above or in these two papers: https://arxiv.org/pdf/1910.01137.pdf , https://arxiv.org/pdf/1904.04266.pdf). 

Note that in file entropy.py in function bipartite_entanglement_entropy(), I get rid of zero eigenvalues of the reduced density matrix before computing the entanglement entropy. The reason for this is, as I said, x*ln(x) -> 0 at x->0.  Calculating x*ln(x) explicitly might be the reason for numerical errors, but I am not sure.

Let me know if something is not clear in the code or you have problems running it.

**Comment by weinbe58** [comment]
Hi @Alehud 

So as far as I can tell it's not an issue with calculating the VN entropy explicitly. You can check this by having the ent_entropy function return the eigenvalues of rdm_A and calculating it for yourself correcting for the small values and this makes little difference. What I did try was to compare the eigenvectors you get when diagonalizing with float32 and float64. I found that, despite the matrices and eigenvalues being exactly the same (within machine precision for float32), the eigenvectors are different between the two precisions meaning that it is possible that the eigensolver is picking different orthogonalization in the different degenerate subspaces depending on the precision of the matrix. Another possibility is that this result may be related to the physics, for example if there are energy levels that have separations smaller than 10^(-7) then float32 may not have enough precision to resolve the energy spacing and the solver is assuming those eigenvalues are degenerate when they are not actually degenerate. 

Hope this helps. 

Phil

**Comment by weinbe58** [comment]
If this is the case then I am confident that is the reason why the entanglement entropy is showing this behavior.

**Comment by Alehud** [comment]
Why would the choice of the eigenvectors matter? Choosing a different eigenbasis in a degenerate subspace should not change the eigenvalues. Or am I missing something?

**Comment by mgbukov** [comment]
I think the point is that the degeneracy in `H` is broken by the numerical precision; this results in some entries of the e'vectors corresponding to this degenerate subspace being slightly off their true values. Then, the partial trace multiplies this error and eventually we get incorrect results in the end.

E.g., for `dtype=np.float32`, the last few values of the e'energies read as

```
3.2360666   3.236067    3.2360675   3.2360675
  3.2360675   3.2360675   3.2360675   3.236068    3.236068    3.236068
  3.236068    3.236068    3.2360687   3.2360687   3.2360687   3.2360687
  3.2360687   3.2360687
```
I tried a few different diagonalization solvers but they all behave similarly:
```
from scipy.linalg import eigh
H_array=H.toarray()
E, V = eigh(H_array, driver='evr', ) # 'ev', 'evd', 'evr', 'evx'
```

On a separate note, here's the essence of the inner workings of quspin that computes the vN entropy (i.e. all checks, etc. striped out):

```
# prepare shapes
sub_sys_A_shifted = tuple(j+1 for j in sub_sys)
sub_sys_B_shifted = tuple(j+1 for j in range(N_2d) if j not in sub_sys)
reordering=(0,)+sub_sys_A_shifted+sub_sys_B_shifted
sub_sizes=(-1, 2**len(sub_sys), 2**(N_2d - len(sub_sys)))

# reshape states
states=(V.T).reshape((-1,)+(2,)*N_2d) 
states=states.transpose(reordering)
states=states.reshape(sub_sizes)

# partial trace
rdms = np.einsum('ijk,ilk->ijl', states, states.conj() ) 

# diagonalize
lmbdas = np.linalg.eigvalsh(rdms)

#
Sent = - np.einsum('ij,ij->i', lmbdas, np.log(lmbdas+np.finfo(lmbdas.dtype).eps) )
```

There aren't too many lines of nontrivial code so we might be able to locate which function gives the problem; my guess is that this happens even before we get to these lines, i.e. in the line `E, V = H.eigh()`. @Alehud one check you can do is to take the e'vectors from your code in all sectors, order them according to ascending e'values to create a new array `V`, and run it thru the above lines to see if Sent comes out correct. 

And finally, note that the 0*log(0) is not the issue when adding the shift to avoid machine precision problems in the log: `np.finfo(lmbdas.dtype).eps`. You can independently verify this behavior: 

```
A=np.array([0.0],dtype=np.float64)
S=-A*np.log(A+np.finfo(A.dtype).eps) # -0.0
```

**Comment by Alehud** [comment]
@mgbukov: the paper is already on the arXiv (https://arxiv.org/abs/2107.09690).

Below, I've written a piece of code that constructs the Hamiltonian using quspin, then uses `networkx` to extract information about the blocks, then solves the eigensystem and calculates the entanglement entropy for each block separately. 

Here, the key difference with how quspin calculates the entanglement entropy is that instead of storing the full eigenvector with a lot of zeros, my code stores only eigenvector components from the block, in addition to which exactly states comprise this block (since this information is relevant for S_ent calculation). This potentially allows for treatment of systems with large full Hilbert space that is block-diagonalized in many relatively small blocks. Maybe you will find this useful if you decide to implement something like this into quspin.

Here's the code: [Sent_from_blocks.txt](https://github.com/weinbe58/QuSpin/files/7071063/Sent_from_blocks.txt)
as well as a clarifying pdf: [document.pdf](https://github.com/weinbe58/QuSpin/files/7071050/document.pdf)

The code outputs the desired plot. 
<img src="https://user-images.githubusercontent.com/35610798/131224816-f33e4776-4f0e-4dad-b15a-d3d816939820.png" width="60%" height="60%">

P.S. If you watch closely, this plot does not precisely match the plot from the beginning of the thread - there is an additional dot at the energy -1 in this one. This is, again, due to the fact that some blocks can have degenerate eigenspectra, and the entanglement entropy depends on which basis in this subspace we choose. For example, some blocks have a Hamiltonian
```
H = [[0, 1, 1],
     [1, 0, 1],
     [1, 1, 0]]
```
whose eigenvalues are 2, -1, -1. So, there is a 2-dimensional degenerate eigenspace, and S_ent is different for different choices of the basis in it. This makes sense mathematically, but I am slightly puzzled about what this means physically.

**Comment by mgbukov** [comment]
@Alehud I re-wrote parts of the example to make use of quspin as much as possible; the modification allows to do the procedure in a specific symmetry sector; take a look. In particular, this removes the need for the extra lines of custom code to compute Sent, and allows us to do this in the presence of symmetries. 

[example27.txt](https://github.com/weinbe58/QuSpin/files/7072303/example27.txt)

Can you send us a short description with some physics motivation and (if you want) a reference to your arXiv preprint -- see this example file where u can see the way to use latex in rst files (txt file format below is only due to github), or any other [quspin example](https://weinbe58.github.io/QuSpin/Examples.html). Or, I can add the reference to the paper myself if you don't know the rst command, just denote the place you want it to appear:

[example27rst.txt](https://github.com/weinbe58/QuSpin/files/7072311/example27rst.txt)

There are more [rst files](https://github.com/weinbe58/QuSpin/tree/master/sphinx/source/examples) for the quspin examples, in case you want to take a look.


About the leftover degeneracy in the block: I think ultimately what this tells you is that the question you ask (plot Sent vs E) is not well-posed, b/c of the reasons we discussed above. In the specific example above, the 3x3 matrix can be further reduced if working in a symmetry subspace which will change the values of Sent. 

What I find interesting is that the blocks you get from fragmentation are not strict sub-blocks of the symmetry blocks, i.e. in a single fragmentation block you can find states belonging to different symmetry sectors. This means that a disambiguous answer to your question in the PS above does not exist (or depends on further conventions, etc.), since if you first apply a symmetry, and then fragmentation you get different Sent values than vice-versa.  But then what is more physical -- the symmetry block or the fragmentation block?!

Because of that, maybe plotting Sent is didactically not the best thing we can show to users (it raises more Q's than it answers), but I leave this to you. If you decide to stick with Sent, pls make a note about this behavior in the rst file, so the users are warned ahead of time ;)

**Comment by Alehud** [comment]
@mgbukov 
> What I find interesting is that the blocks you get from fragmentation are not strict sub-blocks of the symmetry blocks, i.e. in a single fragmentation block you can find states belonging to different symmetry sectors.

This doesn't make sense to me. A fragmentation block cannot be comprised of states from different symmetry sectors, since states from different symmetry sectors by the very definition of symmetry cannot have non-zero matrix elements between them.
If you have a symmetry, you have a corresponding block in the Hamiltonian, meaning that if you choose a basis state of this block (i.e. a row of H), it can have non-zero elements _only_ inside of this block (otherwise it's not a symmetry). The only thing that can happen is that fragmentation blocks lie fully inside the symmetry blocks, i.e. symmetry blocks are further block-diagonalized.

Also, what made you think like this? Have you tried larger lattices of this model? Because, as I said before, on the 3x3 lattice this model has no fragmentation at all. All blocks are due to symmetries.

**Comment by mgbukov** [comment]
I did the following quick check (maybe I missed something, so make sure to double-check this): I first looked at the values of Sent vs E w/o symmetries; after adding Nup, the plot did not change which is good. But when I added the translation symmetries, I saw that the max value for Sent (over all E) was bigger than w/o the transl symmetry, while the energies E in the momentum block were a subset of the total block (suggesting the transl symmetry is indeed present). This maximum value (about Sent=1.74) occurs in a block of 2 states where the energies are non-degenerate (E=+/-1). 

So unless the Hamiltonian has no translation symmetry (quspin doesn't do checks for the general basis classes), the wrong values of Sent should arise due to using the momentum blocks.

[example27py.txt](https://github.com/weinbe58/QuSpin/files/7072841/example27py.txt)

**Comment by mgbukov** [comment]
ok I found a few bugs with using networkx; but the question remains after fixing them: I get the correct energies, but not the correct entropies:

[example27py.txt](https://github.com/weinbe58/QuSpin/files/7072903/example27py.txt)

**Comment by Alehud** [comment]
@mgbukov The smallest blocks in this model on the 3x3 lattice originate from the magnetization conservation in each row and column. This symmetry does not commute with translation in the perpendicular direction (nor with other spatial symmetries or spin inversion). In my understanding, one has to choose a commuting set of symmetries to work with.

In this case, I think, conserved row and column magnetization is the most convenient and most full set of commuting symmetries. (of course the total magnetization is conserved, as well as the total dipole moment, but they are just linear combinations of row and column magnetizations)

**Comment by Alehud** [comment]
Regarding your code in `example27.txt`: in order to use quspin for entanglement entropy calculation you had to pad the eigenvectors with zeros
```
V_full = np.zeros((basis.Ns, len(block_basis_states)), dtype=V.dtype)
```
which I feel like an unnecessary storage of many zeros. It also seems to slow down the calculation. For example, set Lx = Ly = 4. On my laptop, your code calculates it in about 10 min, while my code in `Sent_from_blocks.txt` calculates it in under 1 min. Additionally, I am pretty sure your code will run into a memory problem for Lx = Ly = 5, since in each eigenvector you will have to store around 2^25 numbers (128 Mb for floats or 256 Mb for doubles), most of which are zeros. The largest block for the Lx = Ly = 5 lattice is of size approximately 2000x2000. Then you will be required to store 128*2000 Mb = 250 Gb of floats or 500 Gb of doubles in RAM. With my code, I managed to calculate Lx = Ly = 5 on a computational cluster with under 64 Gb of RAM.

This is why I think that for Hamiltonians that are extensively block diagonalized my approach (i.e., instead of storing an eigenvector explicitly, only store its non-zero components and basis states corresponding to these components) might allow computation of system sizes that are inaccessible for "honest" S_ent calculation. This might be a useful flag for quspin's `ent_entropy(...)` function that switches between two modes of calculating S_ent.

**Comment by mgbukov** [comment]
@Alehud about the symmetries: I am still not completely happy with your argument above. I don't think the problem is with the non-commutativity of the symmetries. Suppose I turn off all symmetries and I only leave the translation symmetry in the basis constructor -- I get the correct e'energies but wrong Sent values. 

How do you define magnetization conservation? If that's $\sum _j S^z$, then this has to commute with any lattice symmetry by definition (and I see that the e'values of `H` with translation+magn symm on are a subset of the e'values with only magn symmetry only). 

About the calculation of Sent: I agree that not having to store a bunch of zeroes is more efficient; if we talk about efficiency, another bottleneck we hit will be building the Hamiltonian. Note that this operation is not needed for the present example (you can apply the operator "+--+" on the fly, i.e. w/o constructing the matrix). But let's resolve the symmetry issue on the 3x3 lattice before we decide on this.

Does your method of computing Sent work in the presence of symmetries? E.g. suppose we don't do the fragmentation blocks at all, and we only consider say translation symmetry: if it gives us the correct values of Sent, then indeed adding a flag in Sent can be useful because it won't compromise the generality of `basis.ent_entropy()`; otherwise we can think about adding a special function in `tools` or so.

**Comment by Alehud** [comment]
@mgbukov
First, about the magnetization conservation: the operators I mean are $\sum_{j in a single row} S^z_j$ and $\sum_{j in a single column} S^z_j$ (Eq.(2) in my paper https://arxiv.org/abs/2107.09690). This is magnetization conservation in _each single_ row and _each single_ column. There are Lx+Ly-1 independent symmetry operators like this. Let me call them $M^row_j, j=1,...,Ly$ and $M_col_i, i=1,...,Lx$. These operators do not commute with spatial symmetries. Intuitively you can think of it like this: measuring magnetization in column i and then translating along x is not the same as first translating along x and then measuring magnetization in column i (in the second case you would measure magnetization in column i-1).

> Suppose I turn off all symmetries and I only leave the translation symmetry in the basis constructor -- I get the correct e'energies but wrong Sent values.

Yes, I don't find it surprising. If you leave only the translation symmetry, then quspin will work in the eigenbasis of the translation operator T, right? Then it will be calculating S_ent of the eigenvectors that are simultaneous eigenvectors of H and of T, but not the eigenvectors of M^row_j, M^col_i. And as we have already understood, the choice of the eigenbasis in degenerate eigenspaces matters for the values of S_ent. Now, the question is: can we not find a common eigenbasis for everything, H, T and M^row_j, M^col_i. The answer is: no, because M^row_j, M^col_i do not commute with T.

**Comment by mgbukov** [comment]
@Alehud Good -- I think we're now on the same page: what we seem to agree on is that the Hamiltonian commutes with translation (so this is one of several lattice symmetries) which block-diagonalizes the Hamiltonian. Fractionalization also block-diagonalizes the Hamiltonian, yet in a different way. Importantly, if we first go to the momentum basis, and then do fractionalization, we get different e'states and thus also different values for the entanglement entropy (than in the original z-basis), because the momentum states are nontrivial superpositions, and we know this changes the entanglement. 

So back to your question about the residual degeneracy in the fractionalization blocks (w/o using symmetries) which leads to Sent not being uniquely defined on that subspace -- I think these extra degeneracies might be due to a leftover symmetry, like translation or diagonal-reflection or spin-flip. If you resolve the degeneracies using translation, then some values of Sent may also change. That's why I was suggesting that we discuss this in the rst file, so users don't get confused when they start playing with the example. 


On a separate note, translation (only k=0,\pi quantum numbers) commutes with parity/reflection and with spin inversion so they can all be used simultaneously in the basis to get a further reduction in the H-space. In general, smaller-sizes H-symmetry-spaces will result in smaller graphs (to find the connected subgraphs of) which would allow you to reach even larger system sizes using fractionalization than without symmetries.   


Now we have to just finish the example: does your method for computing Sent also apply to bases with symmetries? I guess it might but you have to identify the correct product bases $|\phi_{A,B}\rangle$ and I think this is cumbersome for the case of lattice symmetries, as compared to working in the z-basis. Currently, `basis.ent_entropy()` has to first lift the state to the full basis, and then it performs the partial trace. Of course, we can say we don't wanna use any symmetry other than particle/magnetization conservation where this is not an issue, yet then the question is why we would suggest people to use quspin for this specific problem in the first place.

**Comment by Alehud** [comment]
@mgbukov
I suppose by "fractionalization" you mean identifying blocks without using any knowledge of symmetries? (let's call it fragmentation)
> does your method for computing Sent also apply to bases with symmetries? I guess it might but you have to identify the correct product bases $|\phi_{A,B}\rangle$ and I think this is cumbersome for the case of lattice symmetries, as compared to working in the z-basis.

I think you are right. One can always first use some symmetry, take a particular block of this symmetry and then run the "fragmentation" procedure inside of this block. But the basis for this block should be comprised of the eigenstates of the symmetry operator. And while in the case of magnetization conservation these eigenstates are simply z-states, in case of translation or other symmetry they are more complicated, and one has to perform a change of basis first. I will try to write some code that does this.

However, I want to emphasize again, that from the very start one has to choose a commuting set of symmetries and strictly stick
only to them. If a model has two symmetries, A and B, such that <img src="https://render.githubusercontent.com/render/math?math=[H,A] = [H,B] = 0">, but they don't commute, <img src="https://render.githubusercontent.com/render/math?math=[A,B] \neq 0">, then we cannot first take symmetry A, block diagonalize H with respect to it, look inside of one of the blocks, and then suddenly try to reduce the basis even more using symmetry B. This will not work, because A and B do not have a common eigenbasis. So, I don't really understand what you mean by this:

> I think these extra degeneracies might be due to a leftover symmetry

I first fix a set of commuting symmetries: <img src="https://render.githubusercontent.com/render/math?math=M^{\text{row}}_j, M^{\text{col}}_i">. I block diagonalize H with respect to them. Now, I cannot further block diagonalize H with respect to any other symmetry, because no other symmetry operator commutes with all of the <img src="https://render.githubusercontent.com/render/math?math=M^{\text{row}}_j, M^{\text{col}}_i">. So, the translation operators <img src="https://render.githubusercontent.com/render/math?math=T^x, T^y"> are now useless, since they are redundant symmetries. In some sense we have already taken care of <img src="https://render.githubusercontent.com/render/math?math=T^x, T^y"> by considering the _full_ set of commuting symmetries. In fact, the translation operators depend on <img src="https://render.githubusercontent.com/render/math?math=M^{\text{row}}_j, M^{\text{col}}_i"> through the relations <img src="https://render.githubusercontent.com/render/math?math=M^{\text{row}}_j T^y = T^y M^{\text{row}}_{j-1}">, <img src="https://render.githubusercontent.com/render/math?math=M^{\text{col}}_i T^x = T^x M^{\text{col}}_{i-1}">. In principle, we can of course choose another set of commuting symmetries and replace one of the <img src="https://render.githubusercontent.com/render/math?math=M^{\text{row}}_j">'s with <img src="https://render.githubusercontent.com/render/math?math=T^y"> (then we also have to take appropriate linear combinations of the remaining <img src="https://render.githubusercontent.com/render/math?math=M^{\text{row}}_j">'s, such that they commute with <img src="https://render.githubusercontent.com/render/math?math=T^y">), and then work in the common eigenbasis of this set of symmetry operators.

So, maybe I didn't get your idea, but I don't understand how we can resolve the remaining degeneracy using spatial symmetries.

**Comment by Alehud** [comment]
@mgbukov 
Is it possible to use numpy arrays inside of the `next_state()` function? How do I declare their types in the `@cfunc` wrapper?

**Comment by mgbukov** [comment]
I haven't seen or done this myself but it might be possible -- you'd have to check numba's documentation.

**Comment by Alehud** [comment]
I have written code that implements row and column particle conservation symmetries.
[row_col_Np.txt](https://github.com/weinbe58/QuSpin/files/7133068/row_col_Np.txt)
Unfortunately, I didn't figure out how to pass numpy arrays to `@cfunc`. So, instead I just declare the array arguments inside of the `next_state()` function. Excuse me this nuisance. Maybe you'll figure out how to do it properly.

Some comments about the code:
These symmetries are represented by the quantum numbers Mx=(Mx[0], ..., Mx[Lx-1]), My=(My[0], ..., My[Ly-1]), which show the number of particles in each column and each row of the (Lx x Ly) square lattice, respectively. (only Lx+Ly-1 symmetries are linearly independent)
Finding binary matrices with predefined row- and column-sums is not a trivial task. There is Ryser's algorithm that constructs one such matrix (or tells that such a matrix doesn't exist), I gave the reference in the code. This is the essence of the function `get_s0_pcon()`. Now, given this initial matrix, I construct the next matrix with quantum numbers Mx, My by iteratively moving particles in each row (in the same manner as it is done by quspin for regular particle conservation) and check if the column sums are preserved. This is how `next_state()` and `get_Ns_pcon()` functions work.

**Comment by Alehud** [comment]
I have also written code with my method of calculating S_ent in the presence of translational symmetry.
[Sent_with_T.txt](https://github.com/weinbe58/QuSpin/files/7133154/Sent_with_T.txt)

First, basis for the Tx=0, Ty=0 block is constructed, and the Hamitlonian for this block. Then, using `networkx`, we 'fragment' it into smaller blocks (what exactly these blocks mean, we don't know; maybe they correspond to some unresolved symmetries or maybe they are due to the fragmentation of the Hilbert space). For each small block, we identify the basis states that span it (eigenstates of Tx, Ty). We want to compute the entanglement entropy of these eigenstates. But the bipartition is defined in real space, so we have to transform the basis into the z-basis. We do it using the transformation matrix S (not the full matrix, but only the submatrix relevant to the current block). We get the eigenstates in the z-basis and compute their entanglement entropy using my method (meaning that we don't store unnecessary zeros at any point in the procedure).

To sum up, it seems to me that my method of computing S_ent is generic and might be useful for situations when one is interested in blocks of the Hamiltonian that are not too large. The method doesn't care about the size of the full Hilbert space, only about the size of the block. This makes it possible to access small blocks in large systems, when the full Hilbert space is intractable.
On the other hand, when the Hamiltonian is block diagonalized into several large blocks, I think your method of computing S_ent is much faster.

**Comment by mgbukov** [comment]
@Alehud thanks, the codes look good, and I was able to run them. 

Can you also prepare the .rst I'd mentioned above files with the physics description and the relevant equations that the users will see online? You have to decide if you want to keep the two examples separately (in which case we'll need one rst file for each), or combine then into a single one (not sure if this makes sense here, but it's up to you).

**Comment by mgbukov** [comment]
I'm closing this issue due to a lack of activity. Feel free to reopen at any time.
